MIT Bitcoin Expo 2021 - Defending Against 99.999% Attacks (#31)
bitcointranscripts/bitcointranscripts: mit-bitcoin-expo/mit-bitcoin-expo-2021/defending-against-99-attack.md

## Utreexo
So there is an application of this which is something I have been working on for a while. If you have a secure accumulator for Bitcoin, rather than just a merkle tree for transactions in a block. Right now in Bitcoin you have a merkle tree for all the transactions in a block in a merkle tree, you take that merkle root and you put that into the block header and you hash it. We can also use this same technology to keep track of every coin in the system instead of in a regular database, we also put this into a merkle tree accumulator. This is something that's working now, it's pretty cool. This is a good application for this sort of reduction of security. Merkle trees rely on collision resistance hash functions. So in general you say okay you add these four aspects, 0,1,2,3 into this set and you hash them together so that you only have this one hash. But then people can prove that 0,1,2,3 were in the original set. 

> So if you want to attack it, you want to insert something then prove something that you didn't insert you need to find two and two prime, where two is not equal to two prime, these two different elements you want to insert where the hashes are the same. You insert one of them then you prove the other one, and the proof for them is the same. And you have broken the system. You've inserted two but then you can prove two prime which is different. This is a collision attack, finding two different things that hash to the same thing. This is easier than a hash preimage attack. If you have a 256 bit hash function, it only takes 2^128 to perform this type of attack. In contrast to a preiamge attackwhere it takes the full 2^256. 

Reviewer: ```suggestion
So if you want to attack it, you want to insert something then prove something that you didn't insert you need to find two and two prime, where two is not equal to two prime, these two different elements you want to insert where the hashes are the same. You insert one of them then you prove the other one, and the proof for them is the same. And you have broken the system. You've inserted two but then you can prove two prime which is different. This is a collision attack, finding two different things that hash to the same thing. This is easier than a hash preimage attack. If you have a 256 bit hash function, it only takes 2^128 to perform this type of attack. In contrast to a preimage attack where it takes the full 2^256. 
```

> So we can sort of mix these, the mining proof of work and prevention of the collision attacks. So you can say you're going to put the block hash confirming a transaction into the UTXO tree. So now you have this commitment to the proof of work. It also commits to its position in the tree so that the attacker can't collide with an arbitrary element. So for example, the hashes are now the number two (the value), the block hash that confirms it and the UTXO. So this makes collision attacks much harder, because instead of performing a single hash to try and find the collision, you need to mine a block for every attempt. And you can't reuse this work, everytime you try one it changes the block itself. You have two merkle trees, one that has the block hash and one that has the UTXO set. Now you've changed the amount of work you have to do. It's still a collision attack, it's still 2^128 operations, but the operation has changed from a hash function to mining an entire block, which currently is something like 2^70 hashes. So that is a huge increase. Even though in some sense its the same equation. 

Reviewer: ```suggestion
So we can sort of mix these, the mining proof of work and prevention of the collision attacks. So you can say you're going to put the block hash confirming a transaction into the UTXO tree. So now you have this commitment to the proof of work. It also commits to its position in the tree so that the attacker can't collide with an arbitrary element. So for example, the hashes are now the number two (the value), the block hash that confirms it and the UTXO. So this makes collision attacks much harder, because instead of performing a single hash to try and find the collision, you need to mine a block for every attempt. And you can't reuse this work, every time you try one it changes the block itself. You have two merkle trees, one that has the block hash and one that has the UTXO set. Now you've changed the amount of work you have to do. It's still a collision attack, it's still 2^128 operations, but the operation has changed from a hash function to mining an entire block, which currently is something like 2^70 hashes. So that is a huge increase. Even though in some sense its the same equation. 
```

> So if it's 32 byte hashes, you'd need to mine 2^128 blocks to collide, or to do a preimgae attack you'de need to do 2^256 hashes. If you reduced your hash to 16 bytes, you'd need to mine 2^64 blocks to collide, or 2^128 hahes for a preimage attack. And 2^64 blocks is a lot of blocks, 18 quintillion blocks. So it does seem like yeah, it's a reduction in security, but really are you meaningfully defending against an attacker that can mine 18 quintillion blocks? 

Reviewer: ```suggestion
So if it's 32 byte hashes, you'd need to mine 2^128 blocks to collide, or to do a preimage attack you'de need to do 2^256 hashes. If you reduced your hash to 16 bytes, you'd need to mine 2^64 blocks to collide, or 2^128 hahes for a preimage attack. And 2^64 blocks is a lot of blocks, 18 quintillion blocks. So it does seem like yeah, it's a reduction in security, but really are you meaningfully defending against an attacker that can mine 18 quintillion blocks? 
```

> So that's the general idea here. Do we need to reduce hash sizes for these things? No. Is too much security a waste? It can be. It can be very efficient to reduce the size of these things by half. To reduce downloads, and reduce disk storgage by half, and speed things up. That's a meaningful gain. Maybe addresses are a little longer than they need to be, but they're not twice as long as they need to be. This is an interesting idea, so it's something I want to put out there. It's something we should thinkg about. Maybe there are other places that people are working on where there's proof of work happening. The Bitcoin blockchain keeps going on, doing enormous amounts of work and maybe we're wasting some of that work. Maybe we can use it to increase efficiency. If you're been looking at the news in the last 6 months since Bitcoin's been going up, a lot of people think Bitcoin is really wasteful and it's pointless. I don't think they're right about it. But are we using this proof of work? It's happening. Are we taking full advantage of this? Can we leverage it for more efficiently and more power in this system?

Reviewer: hard to see the changes with this diff
storgage, thinkg, efficiently were changed below

```suggestion
So that's the general idea here. Do we need to reduce hash sizes for these things? No. Is too much security a waste? It can be. It can be very efficient to reduce the size of these things by half. To reduce downloads, and reduce disk storage by half, and speed things up. That's a meaningful gain. Maybe addresses are a little longer than they need to be, but they're not twice as long as they need to be. This is an interesting idea, so it's something I want to put out there. It's something we should think about. Maybe there are other places that people are working on where there's proof of work happening. The Bitcoin blockchain keeps going on, doing enormous amounts of work and maybe we're wasting some of that work. Maybe we can use it to increase efficiency. If you're been looking at the news in the last 6 months since Bitcoin's been going up, a lot of people think Bitcoin is really wasteful and it's pointless. I don't think they're right about it. But are we using this proof of work? It's happening. Are we taking full advantage of this? Can we leverage it for more efficiency and more power in this system?
```

And then I have this meme about hashes. 

## Conclusion 

> Thanks! I've put this out as an idea. I'm not like, hey let's decrease the secuirty of all these things, but I do think it's an interesting topic. Are there areas here where we're leaving performance on the table? Or are we paradoxically reducing security by going too far? Questions, comments, disagreements? 

Reviewer: ```suggestion
Thanks! I've put this out as an idea. I'm not like, hey let's decrease the security of all these things, but I do think it's an interesting topic. Are there areas here where we're leaving performance on the table? Or are we paradoxically reducing security by going too far? Questions, comments, disagreements? 
```

> I want to reiterate something that Neha said at the end of the last presentation. I work at the DCI, if you're interested in this kind of stuff get in touch. Get in contact we're pretty cool. Everything is remote now, so wherever you are in the world it doesn't make much of a difference. Are there any questoins?

Reviewer: ```suggestion
I want to reiterate something that Neha said at the end of the last presentation. I work at the DCI, if you're interested in this kind of stuff get in touch. Get in contact we're pretty cool. Everything is remote now, so wherever you are in the world it doesn't make much of a difference. Are there any questions?
```

Tadge: I didn't go into this in enough detail here but the idea of the utreexo project is this accumulator I'm working on with a bunch of open source contributors. It's sort of like pruning to whole UTXO set. In Bitcoin you can store the whole blockchain which is 256 gigs, or you can prune it and say look once I have the block I don't really need to keep it around. And I can store the UTXO set which is about 4 gigs, and is the current state of who owns what. And this basically prunes that as well. Don't store the UTXO set, store a little merkle root of it. It's not quite a merkle root, but it's very small, it's less than a kilobyte. And then transactions can prove that the UTXO they're spending exist with these merkle proofs. That's one example of where we can potentially leverage smaller proofs. It is a sort of different way of pruning the UTXO set. You don't need the blockchain, you don't need to UTXO set, you can run a full node in a very small amount of space.

Moderator: we have time for about two more questions, the next question is: Converse to what you were saying about reducing the hash size, isn't increasing the hash size supposed to future proof the system instead?

> Tadge: Well, maybe. But you can't. You could say, let's put something new in Bitcoin where we have 512 bit pubkey hashes. Or we use an elliptic curve that 400 or 500 bits long, so we double the security. Well okay, but we still have almost everyone using the previous parts of the system. If somebody makes a computer that can derive a private key from a public key, which takes 2^128 ellitic curve ops. Maybe you're safe, but everyone else isn't, so the whole system isn't. If you're going to make your Bitcoins twice as secure, it doesn't matter, if everybody else's coins have been stolen it doesn't matter because the system has collapsed. It's no use to say "I still have mine", because everybody just leaves the system. So there may be ways to upgrade, but you kind of need everyone to upgrade, or almost everyone. If there's 5 million of the 21 million coins are stealable, the system is in trouble. 

Reviewer: ```suggestion
Tadge: Well, maybe. But you can't. You could say, let's put something new in Bitcoin where we have 512 bit pubkey hashes. Or we use an elliptic curve that 400 or 500 bits long, so we double the security. Well okay, but we still have almost everyone using the previous parts of the system. If somebody makes a computer that can derive a private key from a public key, which takes 2^128 elliptic curve ops. Maybe you're safe, but everyone else isn't, so the whole system isn't. If you're going to make your Bitcoins twice as secure, it doesn't matter, if everybody else's coins have been stolen it doesn't matter because the system has collapsed. It's no use to say "I still have mine", because everybody just leaves the system. So there may be ways to upgrade, but you kind of need everyone to upgrade, or almost everyone. If there's 5 million of the 21 million coins are stealable, the system is in trouble. 
```

Moderator: What about quantum computers, is Bitcoin ready for that? 

> Tadge: Right now no, definitely not. If you have a quantum computer, you can grab all of Satoshi's coins, there's so many unhashed public keys that are known that havve enormous amounts of coins so you just pick one and perform a quantum operation. You perform that attack and you take all the coins. Right now it's not, it could be. A system like Bitcoin would be pretty amenable to being quantum safe. There are a bunch of quantum safe signature schemes that we could use, but they're worse in that they're bigger and slower and stuff. But you could do it. The question of upgrading is very difficult. You sort of have to change everybody's keys. Even if it's just a quarter of the coins that are stealable, is this system going to survive that? There's also the question of quantum computers and proof of work, how do those interact? So right now no, it's something people are thinking about but it's hard to upgrade everyone. Maybe if a quantum computer comes out or it's imminent, then it's something that's scary enough that nearly everyone will move, and we make some kind of soft fork which says all of these old coins are frozen forever. But it would be very disruptive. We're not really planning for that right now.

Reviewer: havve
```suggestion
Tadge: Right now no, definitely not. If you have a quantum computer, you can grab all of Satoshi's coins, there's so many unhashed public keys that are known that have enormous amounts of coins so you just pick one and perform a quantum operation. You perform that attack and you take all the coins. Right now it's not, it could be. A system like Bitcoin would be pretty amenable to being quantum safe. There are a bunch of quantum safe signature schemes that we could use, but they're worse in that they're bigger and slower and stuff. But you could do it. The question of upgrading is very difficult. You sort of have to change everybody's keys. Even if it's just a quarter of the coins that are stealable, is this system going to survive that? There's also the question of quantum computers and proof of work, how do those interact? So right now no, it's something people are thinking about but it's hard to upgrade everyone. Maybe if a quantum computer comes out or it's imminent, then it's something that's scary enough that nearly everyone will move, and we make some kind of soft fork which says all of these old coins are frozen forever. But it would be very disruptive. We're not really planning for that right now.
```
